### DISCUSSION TITLE
Does working around the runtime limitations of System.Range result in idiomatic C#?

### DISCUSSION THREAD

**Discussion by DavidArno** [dotnet/csharplang#6864] [Language Ideas] [answered]
Not sure this has the best title, but I think it fits. Consider the following code:
```cs
foreach (var x in 1 .. 10)
{
    Console.WriteLine(x);
}
```
Currently, whilst this code looks sensible and one might assume it would work, it results in:
```
error CS1579: foreach statement cannot operate on variables of type 'Range' because 'Range' does not 
contain a public instance or extension definition for 'GetEnumerator'
```
`System.Range` doesn't implement a `GetEnumerator`, nor implement `IEnumerable<int>`. The reasons why that might be aren't relevant here as I'm not interesting in arguing for a runtime change. That's a topic for `dotnet/runtime`.

What I'm interesting in exploring here is whether the brevity of the range syntax in C# justifies creating an extension to `Range` to allow the above code to work, or whether the gnarly edge cases that ranges can have mean this ends up feeling like an abuse of the language.

**Question 1 - What happens in the case of `1 .. ^10`?**
For the code:
```cs
foreach (var x in 1 .. 10)
{
    Console.Write($"{x} ");
}
Console.WriteLine();
```
I would expect (and I assume most others would to) that the output would be:
```
1 2 3 4 5 6 7 8 9
```
as ranges are "exclusive". Ranges including `^` are effectively inclusive though, so for:
```cs
foreach (var x in 1 .. ^10)
{
    Console.Write($"{x} ");
}
Console.WriteLine();
```
It seems reasonable to expect that the output would be:
```
1 2 3 4 5 6 7 8 9 10
```

**Question 2 - What happens in the case of `10 .. 1`? and `10 ... ^1`?**
Since `10 .. 1` is valid syntax for a range, it seems reasonable that it work in what seems the obvious way:
```cs
foreach (var x in 10 .. 1)
{
    Console.Write($"{x} ");
}
Console.WriteLine();
```
would output:
```
10 9 8 7 6 5 4 3 2
```
And for the "inclusive option":
```cs
foreach (var x in 10 .. ^1)
{
    Console.Write($"{x} ");
}
Console.WriteLine();
```
the output would be:
```
10 9 8 7 6 5 4 3 2 1
```
**Question 3 - What happens with "edge cases" (`^1..10`, `1..`, `..10` and `..`)?**
I'll leave the first to last here as that's the difficult one. 
So taking `1..` first, that is equivalent to `1..^0`. The `Range` type itself doesn't capture that no value was specified, that's purely C# "syntactic sugar". So taking the above rules:
```cs
foreach (var x in 1 ..)
{
    Console.Write($"{x} ");
}
Console.WriteLine();
```
results in the output:
```
1 0
```
The same then applies to `..10` (that's `0..10`, so the output is `0 1 2 3 4 5 6 7 8 9`) and `..` (that's `0..^0`, so the output is `0`).

I'm not sure if these results would pass the POLA test though. They are artefacts of the way that the C# range syntax maps to the `Range` type. So either they are are accepted as such, or they start to feel like abuses of the language. And I'm not sure which it is.

The last case `^1..10` has no obvious solution. It looks like nonsense. The only way I can think to handle it is to just ignore the `^` and view it as equivalent to `1..10`.

A lot of these issues could be addressed if the language team chose to have the language extend `Range` functionality for `foreach` loops, but I doubt there's an appetite for it. So a 3rd party extension that accepts those warts seems a way forward. But I'm still unsure whether this is an abuse of the language and so is best left unwritten.

I'd be interested to know the thoughts of others.

**Comment by CyrusNajmabadi** [comment, marked as answer]
This appears to be a dupe of https://github.com/dotnet/runtime/discussions/80293.  As mentioned in that discussion, this area was already looked into, and the determination was taht the language already exposes the appropriate extensibility point here to allow for this if the owners of this type (the runtime) decide this is desirable for them.  If they don't believe this is appropriate, that's ok as well as users can provide this capability if they want.  From teh language's perspective this is an appropriate separation that needs no extra complexity anywhere. 

In other words, this scenario is already feasible, and control of it is in the right hands from the perspective of the language.  Either the type owners if they want this, or the user if they want it.  There's no desire to add more complexity to enable a scenario that is already enabled today.

**Comment by DavidArno** [comment]
Thanks, @CyrusNajmabadi. 

As no one else has commented, I'll assume others agree with you that the language provides the hooks to allow others to implement an enumerable range if they so desire and that doing so is not seen as an abuse of the syntax.

I appreciate your assistance.

**Comment by HaloFour** [comment]
I am a little disappointed that the conversation around ranges did not extend beyond slicing.  It was stated at the time that other scenarios could be considered later, but I think the design became so slicing-centric that perhaps it'd be difficult to address that now.

**Reply by DavidArno** [reply]
Indeed. The limitation of ranges only being able to take values >= 0 is presumably because that's all that makes sense for slices. But there's no reason why a range itself couldn't be eg `-50 .. 50`, other than the type itself throws if given a negative value. But as Cyrus says, this is now a runtime issue, not a language issue.

**Comment by nuiva** [comment]
> Ranges including `^` are effectively inclusive though, so for:
>
> ```C#
> foreach (var x in 1 .. ^10)
>{
>     Console.Write($"{x} ");
> }
> Console.WriteLine();
> ```
>
> It seems reasonable to expect that the output would be:
> 
> ```
> 1 2 3 4 5 6 7 8 9 10
> ```

As far as I see, `^` signifies "from the end", so `1 .. ^10` should output `1 2 3 4 5 ... 2147483637`, where the last number is `int.MaxValue - 10`. Admittedly, that isn't very useful, but having two different meanings for `^x` sounds bad. Some might be deluded to think that `array[1..^10]` grabs an inclusive slice of 10 elements!
